### PR TITLE
Send status report in IM server side when server encounter the error

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
When server don't have readhandler/subscribe handler, server would send
status report with RESOURCE_EXHAUSTED.
When server don't have writehandler/commandhandler, server would send
status report with BUSY.
For other error generated when recieving incoming request in server,
server would send status report with FAILURE

#### Change overview
See above

#### Testing
Add unit test to overflow the range for read/subscribe handler pool, then check the error

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -343,49 +324,39 @@ CHIP_ERROR InteractionModelEngine::OnReadInitialRequest(Messaging::ExchangeConte
     {
         if (readHandler.IsFree())
         {
-            err = readHandler.Init(mpExchangeMgr, mpDelegate, apExchangeContext, aInteractionType);
-            SuccessOrExit(err);
-            err               = readHandler.OnReadInitialRequest(std::move(aPayload));
-            apExchangeContext = nullptr;
-            break;
+            ReturnErrorOnFailure(readHandler.Init(mpExchangeMgr, mpDelegate, apExchangeContext, aInteractionType));
+            ReturnErrorOnFailure(readHandler.OnReadInitialRequest(std::move(aPayload)));
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:328]: Again, need to figure out who's responsible for sending an error response in which failure cases here.  Followup.

File: src/app/InteractionModelEngine.cpp [hunk 2]
```diff
@@ -343,49 +324,39 @@ CHIP_ERROR InteractionModelEngine::OnReadInitialRequest(Messaging::ExchangeConte
...
+            aStatus = Protocols::InteractionModel::Status::Success;
+            return CHIP_NO_ERROR;
         }
     }
 
-exit:
+    ChipLogProgress(InteractionModel, "no resource for %s interaction",
+                    aInteractionType == ReadHandler::InteractionType::Subscribe ? "Subscribe" : "Read");
+    aStatus = Protocols::InteractionModel::Status::ResourceExhausted;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:336]: This makes sense for a subscribe, but should a regular read just return Busy?  Or is it that too many subscriptions can keep reads from working?  If so, followup: we should reserve some read handlers for reads.

File: src/app/InteractionModelEngine.cpp [hunk 3]
```diff
@@ -343,49 +324,39 @@ CHIP_ERROR InteractionModelEngine::OnReadInitialRequest(Messaging::ExchangeConte
...
     for (auto & writeHandler : mWriteHandlers)
     {
         if (writeHandler.IsFree())
         {
-            err = writeHandler.Init(mpDelegate);
-            SuccessOrExit(err);
-            err               = writeHandler.OnWriteRequest(apExchangeContext, std::move(aPayload));
-            apExchangeContext = nullptr;
-            break;
+            ReturnErrorOnFailure(writeHandler.Init(mpDelegate));
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:351]: Again, something here should be sending error responses?  Followup.

File: src/app/InteractionModelEngine.cpp [hunk 4]
```diff
@@ -421,31 +391,46 @@ CHIP_ERROR InteractionModelEngine::OnUnsolicitedReportData(Messaging::ExchangeCo
...
     }
+
+exit:
+    if (status != Protocols::InteractionModel::Status::Success)
+    {
+        err = StatusResponse::SendStatusResponse(status, apExchangeContext, false /*aExpectResponse*/);
+        if (err != CHIP_NO_ERROR)
+        {
+            apExchangeContext->Abort();
+        }
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:431]: Why is this bit needed?

We just got a message, we handled it, we did not send a response.  The exchange will close itself as soon as we return from here.  Why are we manually aborting it?

File: src/app/InteractionModelEngine.h [hunk 5]
```diff
@@ -199,10 +197,8 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate, public Comman
 
     void OnDone(CommandHandler & apCommandObj) override;
 
-    CHIP_ERROR OnUnknownMsgType(Messaging::ExchangeContext * apExchangeContext, const PayloadHeader & aPayloadHeader,
-                                System::PacketBufferHandle && aPayload);
     CHIP_ERROR OnInvokeCommandRequest(Messaging::ExchangeContext * apExchangeContext, const PayloadHeader & aPayloadHeader,
-                                      System::PacketBufferHandle && aPayload);
+                                      System::PacketBufferHandle && aPayload, Protocols::InteractionModel::Status & aStatus);
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:201]: The semantics around this aStatus value need to be documented.  It's entirely non-obvious from this (undocumented API) that the caller pre-sets it to failure and the callee is expected to set it to success if it does not want an automatic status response message to be sent.

Same for the other APIs here.

File: src/app/StatusResponse.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+#include <app/MessageDef/StatusResponseMessage.h>
+#include <app/StatusResponse.h>
+
+namespace chip {
+namespace app {
+CHIP_ERROR StatusResponse::SendStatusResponse(Protocols::InteractionModel::Status aStatus,
+                                              Messaging::ExchangeContext * apExchangeContext, bool aExpectResponse)
+{
+    using Protocols::InteractionModel::Status;
```
Comment (bzbarsky-apple) [inline src/app/StatusResponse.cpp:27]: Unused; please remove this line.

File: src/app/StatusResponse.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    StatusResponseMessage::Parser response;
+    System::PacketBufferTLVReader reader;
+    reader.Init(std::move(aPayload));
+    ReturnErrorOnFailure(reader.Next());
+    ReturnErrorOnFailure(response.Init(reader));
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+    ReturnErrorOnFailure(response.CheckSchemaValidity());
+#endif
+    ReturnErrorOnFailure(response.GetStatus(aStatus));
+    ChipLogProgress(InteractionModel, "Receive status response, status is %" PRIu16, to_underlying(aStatus));
```
Comment (bzbarsky-apple) [inline src/app/StatusResponse.cpp:59]:
```suggestion
    ChipLogProgress(InteractionModel, "Received status response, status is %" PRIu16, to_underlying(aStatus));
```

File: src/app/ReadClient.cpp [hunk 8]
```diff
@@ -287,6 +248,22 @@ CHIP_ERROR ReadClient::OnMessageReceived(Messaging::ExchangeContext * apExchange
         mpExchangeCtx = nullptr;
         SuccessOrExit(err);
     }
+    else if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        Protocols::InteractionModel::Status status;
+        VerifyOrExit(apExchangeContext == mpExchangeCtx, err = CHIP_ERROR_INCORRECT_STATE);
+        err           = StatusResponse::ProcessStatusResponse(apExchangeContext, std::move(aPayload), status);
+        mpExchangeCtx = nullptr;
```
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:256]: Why?  The `exit:` block belowe will handle that, since we always end up in an error state here, right?

File: src/app/WriteClient.cpp [hunk 9]
```diff
@@ -277,15 +277,32 @@ CHIP_ERROR WriteClient::OnMessageReceived(Messaging::ExchangeContext * apExchang
...
+    if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::WriteResponse))
+    {
+        err = ProcessWriteResponseMessage(std::move(aPayload));
+        SuccessOrExit(err);
+    }
+    else if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        Protocols::InteractionModel::Status status;
+        err           = StatusResponse::ProcessStatusResponse(apExchangeContext, std::move(aPayload), status);
+        mpExchangeCtx = nullptr;
```
Comment (bzbarsky-apple) [inline src/app/WriteClient.cpp:291]: Why? The `exit:` block handles this.

File: src/app/WriteClient.cpp [hunk 10]
```diff
@@ -277,15 +277,32 @@ CHIP_ERROR WriteClient::OnMessageReceived(Messaging::ExchangeContext * apExchang
...
-    err = ProcessWriteResponseMessage(std::move(aPayload));
+    if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::WriteResponse))
+    {
+        err = ProcessWriteResponseMessage(std::move(aPayload));
+        SuccessOrExit(err);
+    }
+    else if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        Protocols::InteractionModel::Status status;
+        err           = StatusResponse::ProcessStatusResponse(apExchangeContext, std::move(aPayload), status);
```
Comment (bzbarsky-apple) [inline src/app/WriteClient.cpp:290]: It looks to me like we have at least 3 copies of this code, which comes down to "Convert status response to CHIP_ERROR value".  Should we just have a StatusResponse function for doing only that, instead of duplicating this code?

File: src/app/util/basic-types.h [hunk 11]
```diff
@@ -29,7 +29,6 @@
 #include <lib/core/DataModelTypes.h>
 
 namespace chip {
-
```
Comment (bzbarsky-apple) [inline src/app/util/basic-types.h:32]: Why this change?

File: src/app/StatusResponse.cpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    response.Status(aStatus);
+    ReturnErrorOnFailure(response.GetError());
+    ReturnErrorOnFailure(writer.Finalize(&msgBuf));
+    ReturnErrorOnFailure(apExchangeContext->SendMessage(Protocols::InteractionModel::MsgType::StatusResponse, std::move(msgBuf),
+                                                        aExpectResponse ? Messaging::SendMessageFlags::kExpectResponse
+                                                                        : Messaging::SendMessageFlags::kNone));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR StatusResponse::ProcessStatusResponse(Messaging::ExchangeContext * apExchangeContext,
```
Comment (bzbarsky-apple) [inline src/app/StatusResponse.cpp:46]: apExchangeContext arg is unused and doesn't really make sense here, I think....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Meant to request changes...  See above review for things that need fixing.

File: src/app/InteractionModelEngine.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:336]: yep, it is latter,  too many subscriptions can keep reads from working since currently readHandler pool is shared by read and subscribe

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:336]: https://github.com/project-chip/connectedhomeip/issues/11720

File: src/app/InteractionModelEngine.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:432]: agree, fixed ,thanks

File: src/app/StatusResponse.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/app/StatusResponse.cpp:27]: done, thanks

File: src/app/StatusResponse.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/StatusResponse.cpp:59]: Done, thanks

File: src/app/WriteClient.cpp (hunk 9 above)
Comment (yunhanw-google) [reply src/app/WriteClient.cpp:291]: agree, fixed, thanks

File: src/app/util/basic-types.h (hunk 11 above)
Comment (yunhanw-google) [reply src/app/util/basic-types.h:32]: accidental change, removed, thanks

File: src/app/StatusResponse.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/StatusResponse.cpp:46]: agree, removed, thanks

File: src/app/WriteClient.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/app/WriteClient.cpp:290]: agree, done, thanks

File: src/app/ReadClient.cpp (hunk 8 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:256]: agree, removed, thanks

File: src/app/InteractionModelEngine.h (hunk 5 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:201]: updated, thanks

File: src/app/CommandSender.cpp [hunk 13]
```diff
@@ -91,12 +92,31 @@ CHIP_ERROR CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExcha
...
+        err = ProcessInvokeResponse(std::move(aPayload));
+        SuccessOrExit(err);
+    }
+    else if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        Protocols::InteractionModel::Status status;
+        err           = StatusResponse::ProcessStatusResponse(apExchangeContext, std::move(aPayload), status);
+        mpExchangeCtx = nullptr;
+        SuccessOrExit(err);
+        if (status == Protocols::InteractionModel::Status::ResourceExhausted)
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/CommandSender.cpp:107]: agree, updated. it seems for onError for read client and writer client, we also need to add StatusIB as parameter?

File: src/app/InteractionModelEngine.cpp [hunk 14]
```diff
@@ -328,10 +309,10 @@ CHIP_ERROR InteractionModelEngine::OnReadInitialRequest(Messaging::ExchangeConte
             bool keepSubscriptions = true;
             System::PacketBufferTLVReader reader;
             reader.Init(aPayload.Retain());
-            SuccessOrExit(err = reader.Next());
+            ReturnErrorOnFailure(reader.Next());
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:313]: Also struggle on this point. I think if server return failure via StatusReponse, the client would get to know what is going on in server side with best effort, and instead of waiting for some time and generate timeout error.

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:329]: yep, would follow-up

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:329]: https://github.com/project-chip/connectedhomeip/issues/11724

File: src/app/InteractionModelEngine.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:352]: https://github.com/project-chip/connectedhomeip/issues/11724

File: src/app/InteractionModelEngine.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:313]: Right, the spec says that you're supposed to send a status response if you get a malformed message....

[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:313]: ack, thanks

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]
